pandas/core/indexes/multi.py

```diff
@@ -3095,8 +3103,54 @@ def _update_indexer(idxr, indexer=indexer):
         # empty indexer
         if indexer is None:
             return Int64Index([])._ndarray_values
+
+        if need_sort:
+            indexer = self._reorder_indexer(seq, indexer)
+
         return indexer._ndarray_values
 
+    def _reorder_indexer(self, seq, indexer: Int64Index) -> Int64Index :
+        """
+        Reorder an indexer of a MultiIndex (self) so that the label are in the
+        same order as given in seq
+
+        Parameters
+        ----------
+        seq : label/slice/list/mask or a sequence of such
+        indexer: an Int64Index for element of self
+
+        Returns
+        -------
+        indexer : a sorted Int64Index of element of self ordered as seq
+        """
+        n = len(self)
+        keys = tuple()
+        # For each level of the sequence in seq, map the level codes with the
+        # order they appears in a list-like sequence
+        # This mapping is then use to reorder the indexer
+        for i, k in enumerate(seq):
+            if com.is_bool_indexer(k):
+                new_order = np.arange(n)[indexer]
+            elif is_list_like(k):
```

Comment:
Is `k` an array? If so, pass `k.dtype` here.

Reply:
k can be a slice, a mask, a single label or a list.
The only case to really handle here is a list, because it is the only kind of input that can be used in a different order than the level. For the other, it just ensure that it is not put out of order.

Reply:
Most of the time (I mean in the example for multi index in the doc for example), k is a python list of label.